src/couch/src/couch_db_split.erl

```diff
@@ -0,0 +1,489 @@
+    CommitTargets = maps:map(fun(_, #target{} = T) ->
+        commit_docs(T)
+    end, NewState#state.targets),
+    NewState#state{targets = CommitTargets}.
+
+
+% Backwards compatibility clause. Seq trees used to hold #doc_infos at one time
+changes_cb(#doc_info{id = Id}, #state{source_db = Db} = State) ->
+    [FDI = #full_doc_info{}] = couch_db_engine:open_docs(Db, [Id]),
+    changes_cb(FDI, State);
+
+changes_cb(#full_doc_info{id = Id} = FDI, #state{} = State) ->
+    #state{source_db = SourceDb, targets = Targets} = State,
+    {Key, Target} = pick_target(Id, State, Targets),
+    FDI1 = process_fdi(FDI, SourceDb, Target#target.db),
+    MaxBuffer = State#state.max_buffer_size,
+    Target1 = acc_and_flush(FDI1, Target, MaxBuffer, fun commit_docs/1),
+    {ok, State#state{targets = Targets#{Key => Target1}}}.
+
+
+commit_docs(#target{buffer = [], db = Db} = Target) ->
+    Target#target{db = Db};
+
+commit_docs(#target{buffer = FDIs, db = Db} = Target) ->
+    Pairs = [{not_found, FDI} || FDI <- lists:reverse(FDIs)],
+    {ok, Db1} = couch_db_engine:write_doc_infos(Db, Pairs, []),
+    {ok, Db2} = couch_db_engine:commit_data(Db1),
+    Target#target{buffer = [], buffer_size = 0, db = Db2}.
+
+
+process_fdi(FDI, SourceDb, TargetDb) ->
+    #full_doc_info{id = Id, rev_tree = RTree} = FDI,
+    Acc = #racc{id = Id, source_db = SourceDb, target_db = TargetDb},
+    {NewRTree, NewAcc} = couch_key_tree:mapfold(fun revtree_cb/4, Acc, RTree),
+    {Active, External} = total_sizes(NewAcc),
+    FDI#full_doc_info{
+        rev_tree = NewRTree,
+        sizes = #size_info{active = Active, external = External}
+    }.
+
+
+revtree_cb(_Rev, _Leaf, branch, Acc) ->
+    {[], Acc};
+
+revtree_cb({Pos, RevId}, Leaf, leaf, Acc) ->
+    #racc{id = Id, source_db = SourceDb, target_db = TargetDb} = Acc,
+    #leaf{deleted = Deleted, ptr = Ptr} = Leaf,
+    Doc0 = #doc{
+        id = Id,
+        revs = {Pos, [RevId]},
+        deleted = Deleted,
+        body = Ptr
+    },
+    Doc1 = couch_db_engine:read_doc_body(SourceDb, Doc0),
+    #doc{body = Body, atts = AttInfos0} = Doc1,
+    External = if not is_binary(Body) -> ?term_size(body); true ->
+        couch_compress:uncompressed_size(Body)
+    end,
+    AttInfos = if not is_binary(AttInfos0) -> AttInfos0; true ->
+        couch_compress:decompress(AttInfos0)
+    end,
+    Atts = [process_attachment(Att, SourceDb, TargetDb) || Att <- AttInfos],
+    Doc2 = Doc1#doc{atts = Atts},
+    Doc3 = couch_db_engine:serialize_doc(TargetDb, Doc2),
+    {ok, Doc4, Active} = couch_db_engine:write_doc_body(TargetDb, Doc3),
+    AttSizes = [{element(3, A), element(4, A)} || A <- Atts],
+    NewLeaf = Leaf#leaf{
+        ptr = Doc4#doc.body,
+        sizes = #size_info{active = Active, external = External},
+        atts = AttSizes
+    },
+    {NewLeaf, add_sizes(Active, External, AttSizes, Acc)}.
+
+
+% This is copied almost verbatim from the compactor
```

Comment:
The logic is the same but there are some minor differences. In the compactor we use couch_file:pread_term here we use read_doc_body. Then since the compactor is implemented inside a PSE engine (so it's under the couch_db_engine API) we'd have to share code across the PSE plugin API boundary and so think it might be better to just keep them separate. The logic also is not terribly convoluted, it's basically open read stream, open write stream, copy, close both and checksum.